x/concentrated-liquidity/simulation/sim_msgs.go

```diff
@@ -1,17 +1,119 @@
 package simulation
 
-// func RandomMsgCreateConcentratedPool(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*clmodeltypes.MsgCreateConcentratedPool, error) {
-// 	return nil, nil
-// }
+import (
+	"errors"
+	"fmt"
+	"math/rand"
 
-// func RandMsgCreatePosition(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*cltypes.MsgCreatePosition, error) {
-// 	return nil, nil
-// }
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	legacysimulationtype "github.com/cosmos/cosmos-sdk/types/simulation"
+
+	osmosimtypes "github.com/osmosis-labs/osmosis/v14/simulation/simtypes"
+	clkeeper "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity"
+	clmodeltypes "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity/model"
+	cltypes "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity/types"
+)
+
+var PoolCreationFee = sdk.NewInt64Coin("stake", 10_000_000)
+
+func RandomMsgCreateConcentratedPool(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*clmodeltypes.MsgCreateConcentratedPool, error) {
+	// generate random values from -13 to 1 (accepted range: -12 to -1)
+	exponentAtPriceOne := sdk.NewInt(rand.Int63n(1+13) - 13)
+	authorizedTickSpacing := []uint64{1, 10, 60, 200}
+
+	// find an address with two or more distinct denoms in their wallet
+	sender, senderExists := sim.RandomSimAccountWithConstraint(createPoolRestriction(k, sim, ctx))
+	if !senderExists {
+		return nil, fmt.Errorf("no sender with two different denoms & pool creation fee exists")
+	}
+
+	// generate 3 coins, use 2 to create pool and 1 for fees. ["stake" denom - contiains invalid metadata]
+	poolCoins, ok := sim.GetRandSubsetOfKDenoms(ctx, sender, 3)
```

Comment:
Why are we getting subset of 3 denoms when `createPoolRestriction` returns accounts with only 2 coins?

Reply:
Hrmm, I think @stackman27 's original intent on this so that we can have 2 denoms for the pool, and then another denom for the pool creation fee, but as @p0mvn mentioned,  I think we should be having 2 rand subset here

Reply:
also the other reason was that "stake" denom fails this check. So i had to use 2 other gamm denoms here 
```
	_, denomExists := server.keeper.bankKeeper.GetDenomMetaData(ctx, msg.Denom0)
	if !denomExists {
		return nil, fmt.Errorf("received denom0 with invalid metadata: %s", msg.Denom0)
	}
```

Reply:
In that case, it seems that here we're taking the first two given coins 
```
denom0 := poolCoins[0].Denom
	denom1 := poolCoins[1].Denom
```

If the case is both 1 and 2 should we not have some sort of logic that verifies that it is not stake denom?

Reply:
Regarding point number 1, the reason it's happening is because of this line `https://github.com/osmosis-labs/osmosis/pull/4094/files#diff-f01e8795f0fb3ac50fd61ff5441aac0f1e4153629f16c05da9c3a8785c408dc9R64`. We can simply change that to be equal to 2 not greater than 2 wihtin the method

Reply:
IMO that should also go in the `createPoolRestriction` as it serves our purpose of createPoolRestrictions

Reply:
I kept it outside because it requires sender which we can only generate with constraint

Reply:
I was rather about doing the following 
```
// createPoolRestriction creates specific restriction for the creation of a pool.
func createPoolRestriction(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) osmosimtypes.SimAccountConstraint {
	return func(acc legacysimulationtype.Account) bool {
		accCoins := sim.BankKeeper().SpendableCoins(ctx, acc.Address)
		hasTwoCoins := len(accCoins) == 2
		
		isStakeDenom := false
		for _, coin := range accCoins {
			if coin.Denom == "stake" {
				isStakeDenom = true
			}
		}
		hasPoolCreationFee := accCoins.AmountOf(PoolCreationFee.Denom).GT(PoolCreationFee.Amount)
		return hasTwoCoins && hasPoolCreationFee && !isStakeDenom
	}
}
```

Suggested this way because having another layer of extra code here (using GetRandSubsetOfKDenoms) seemed unnecessary when we're already calling a method that does this role

Reply:
the issue with this is that we have to have stake denom because that's being used as CreationFee, also any tokens being minted from accCoins contains "stake" denom, but in our case we cannot use stake to create Pool because of the "invalid metadata" problem

Reply:
Can we look into registering stake in bank to get past the invalid metadata problem

Reply:
Yea we can do that by using `SetDenomMetaData` which we'll have to import from bankKeeper, which will break incentives, superfluid simulator. I didn't wanted this to be a big diff so i found this loophole

Reply:
tbh i think we should just rewrite the rest of module simulator to follow gamm structure

Reply:
I tried many things such as;
1. change feeDenom from "stake" to "uosmo" --> Error because: "uosmo" has invalid metadata + user doesnot contain the denom 
2. Set denom metadata before test --> Error: `SetDenomMetaData` cannot be added to simtypes.bankKeeper without breaking incentives, superfluid.  Currently, we can only access specific bankKeeper method in simulator, I created an issue to fix this https://github.com/osmosis-labs/osmosis/issues/4184